lib/matplotlib/artist.py

```diff
@@ -720,9 +720,20 @@ def set_path_effects(self, path_effects):
     def get_path_effects(self):
         return self._path_effects
 
-    def get_figure(self):
-        """Return the `.Figure` instance the artist belongs to."""
-        return self.figure
+    def get_figure(self, root=False):
+        """
+        Return the `.Figure` or `.SubFigure` instance the artist belongs to.
+
+        Parameters
+        ----------
+        root : bool, default=False
+            If False, return the (Sub)Figure this artist is on.  If True,
+            return the root Figure for a nested tree of SubFigures.
+        """
+        if root:
+            return self._figure._figure
```

Comment:
Does this hold in general, i.e. also for deeper subfigure nestings?

Reply:
My test has the axes on the second subfigure down.

Reply:
Ok, I now see that this is technically correct: From https://github.com/matplotlib/matplotlib/issues/28170#issuecomment-2094524856

> [for subfigures] I explicitly set `self.figure =  parent.figure` which therefore resolves to the root figure, and I think that is because artists have things like `self.figure.stale` and `self.figure.suppressComposite` which would somehow have to magically thread themselves through.

But still needs some explanation, because it looks confusing: `self._figure` is the enclosing (sub)figure. and the `_figure` attribute of that is always the root figure. Maybe it helps to rename `Artist._figure` to `Artist._enclosing_figure`. Also, the second access is to a private attribute of (sub)figures, which general artists have no business with - we're relying on an implementation detail of a different class. Possibly a better spelling would be
```suggestion
            return self._enclosing_figure.get_figure(root=True)
```

Reply:
I'd go with `_root_figure` to make the tree structure clear.

Reply:
Oh OK, it never occurred to me that we shouldn't use private attributes on another class in the same package.  But that means we shouldn't use `self._enclosing_figure._root_figure` so @jklymak have I misunderstood your comment?

Reply:
Yeah, I thought you meant the `_root_figure` by `_enclosing_figure`.  Maybe call the artists' subfigure the `_parent_figure` or `_parent_subfigure` if we want to be more specific.  I'd keep with standard tree terminology?  https://en.wikipedia.org/wiki/Tree_(data_structure)#:~:text=A%20node%20that%20has%20a,same%20parent%20are%20sibling%20nodes.

Reply:
I think `_parent_figure` as it may be either a `Figure` or a `SubFigure`.  I note also that we also already have `SubFigure.parent`, so it's good to be consistent with that.